src/neuroconv/tools/nwb_helpers/_metadata_and_file_helpers.py

```diff
@@ -194,53 +194,63 @@ def make_or_load_nwbfile(
+        if append_mode:
             load_kwargs.update(mode="r+", load_namespaces=True)
+
+            # Check if the selected backend is the backend of the file in nwfile_path
+            backends_that_can_read = [
+                backend_name
+                for backend_name, backend_io_class in BACKEND_NWB_IO.items()
+                if backend_io_class.can_read(path=str(nwbfile_path_in))
+            ]
+            # Future-proofing: raise an error if more than one backend can read the file
+            assert (
+                len(backends_that_can_read) <= 1
+            ), "More than one backend is capable of reading the file! Please raise an issue describing your file."
+            if backend not in backends_that_can_read:
+                raise IOError(
+                    f"The chosen backend ('{backend}') is unable to read the file! "
+                    f"Please select '{backends_that_can_read[0]}' instead."
+                )
         else:
             load_kwargs.update(mode="w")
 
-        backends_that_can_read = [
-            backend_name
-            for backend_name, backend_io_class in BACKEND_NWB_IO.items()
-            if backend_io_class.can_read(path=str(nwbfile_path_in))
-        ]
-        # Future-proofing: raise an error if more than one backend can read the file
-        assert (
-            len(backends_that_can_read) <= 1
-        ), "More than one backend is capable of reading the file! Please raise an issue describing your file."
-        if load_kwargs["mode"] == "r+" and backend not in backends_that_can_read:
-            raise IOError(
-                f"The chosen backend ('{backend}') is unable to read the file! "
-                f"Please select '{backends_that_can_read[0]}' instead."
-            )
-
         io = backend_io_class(**load_kwargs)
 
+    nwbfile_loaded_succesfully = True
     try:
         if load_kwargs.get("mode", "") == "r+":
             nwbfile = io.read()
         elif nwbfile is None:
             nwbfile = make_nwbfile_from_metadata(metadata=metadata)
         yield nwbfile
-    except Exception as e:
-        success = False
-        raise e
+    except Exception as load_error:
+        nwbfile_loaded_succesfully = False
+        raise load_error
     finally:
         if nwbfile_path_in:
+            nwbfile_written_succesfully = True
             try:
-                if success:
+                if nwbfile_loaded_succesfully:
                     io.write(nwbfile)
 
                     if verbose:
                         print(f"NWB file saved at {nwbfile_path_in}!")
+            except Exception as write_error:
+                nwbfile_written_succesfully = False
+                raise write_error
             finally:
                 io.close()
 
-                if not success and not file_initially_exists:
+                if not nwbfile_loaded_succesfully and not file_initially_exists:  # Not sure we need this
```

Comment:
Should we moreve this as you suggested? I  don't see how we ever get here.

Reply:
This is required because apparently the file is created by the IO and not by write (as I thought):

```python
from pynwb import NWBHDF5IO

nwb_file_path = Path("./test.nwb")
nwb_file_path.unlink(missing_ok=True)
assert not nwb_file_path.exists()
NWBHDF5IO(path=nwb_file_path, mode="w")
assert nwb_file_path.exists()
```

Reply:
Nice catch; I expanded the logic for the one cleanup condition and made it more explicit about when we'd actually want to cleanup